experimental/codepointtrie/src/codepointtrie.rs

```diff
@@ -0,0 +1,258 @@
+    const ENUM_VALUE: ValueWidthEnum = ValueWidthEnum::Bits32;
+
+    fn cast_to_widest(self) -> u32 {
+        self
+    }
+}
+
+// TrieType trait
+
+pub trait TrieType {
+    const FAST_MAX: u32;
+    const ENUM_VALUE: TrieTypeEnum;
+}
+
+pub struct Fast;
+impl TrieType for Fast {
+    const FAST_MAX: u32 = FAST_TYPE_FAST_INDEXING_MAX;
+    const ENUM_VALUE: TrieTypeEnum = TrieTypeEnum::Fast;
+}
+
+pub struct Small;
+impl TrieType for Small {
+    const FAST_MAX: u32 = SMALL_TYPE_FAST_INDEXING_MAX;
+    const ENUM_VALUE: TrieTypeEnum = TrieTypeEnum::Small;
+}
+
+pub struct CodePointTrie<'trie, W: ValueWidth, T: TrieType> {
+    header: CodePointTrieHeader,
+    index: ZeroVec<'trie, u16>,
+    data: ZeroVec<'trie, W>,
+    _marker_ty: PhantomData<T>,
+}
+
+pub struct CodePointTrieHeader {
+    pub index_length: u32,
+    pub data_length: u32,
+    pub high_start: u32,
+    pub shifted12_high_start: u16,
+    pub index3_null_offset: u16,
+    pub data_null_offset: u32,
+    pub null_value: u32,
+}
+
+// TODO: add Rust-doc that includes examples
+
+impl<'trie, W: ValueWidth, T: TrieType> CodePointTrie<'trie, W, T> {
+    /// Create a new [`CodePointTrie`] backed by borrowed data for the `index`
+    /// array and `data` array.
+    pub fn try_new(
+        header: CodePointTrieHeader,
+        index: ZeroVec<'trie, u16>,
+        data: ZeroVec<'trie, W>,
+    ) -> Result<CodePointTrie<'trie, W, T>, Error> {
+        if header.data_length < ERROR_VALUE_NEG_DATA_OFFSET {
+            return Err(Error::FromDeserialized {
+                reason: "Data array must be large enough to contain error value",
+            });
+        }
+
+        if header.data_length < HIGH_VALUE_NEG_DATA_OFFSET {
+            return Err(Error::FromDeserialized {
+                reason:
+                    "Data array must be large enough to contain value for range highStart..U+10FFFF",
+            });
+        }
+
+        if index.len() as u32 != header.index_length {
+            return Err(Error::FromDeserialized {
+                reason: "Length of index array does not match corresponding header value",
+            });
+        }
+
+        // Index array values are 16-bit width, but in one special case they
+        // are interpreted as 18-bit values. Therefore, the data array length
+        // cannot be more than 2^18.
+        if data.len() as u32 > 262144 {
+            return Err(Error::FromDeserialized {
+                reason: "Length of data array is too large",
+            });
+        }
+
+        if data.len() as u32 != header.data_length {
+            return Err(Error::FromDeserialized {
+                reason: "Length of data array does not match corresponding header value",
+            });
+        }
+
+        // Note: this particular constructor is "templatized" through Rust's
+        // generics type system, so callers to this constructor function must
+        // necessarily know ahead of time which trie type and value width their
```

Comment:
What did y'all end up with as a solution for when the caller does not know/does not care what the type and/or value width is?

Reply:
We have one, it's just not implemented yet.

Reply:
IIUC, this is still necessary for de-/serialization purposes. Manish's sketch of the "inner" and "wrapper" types is still the plan for now, and I came across a Q&A online on the topic that had pretty much the same advice. I'm currently working up to it by starting with tasks for adding serialization to basic structs first.